frame/support/src/traits/hooks.rs

```diff
@@ -63,7 +63,12 @@ pub trait OnFinalize<BlockNumber> {
 ///
 /// Implementing this lets you express what should happen for your pallet before
 /// block finalization (see `on_finalize` hook) in case any remaining weight is left.
-pub trait OnIdle<BlockNumber> {
+pub trait OnIdle<BlockNumber: 'static> {
+	const ON_IDLE_FUNCTIONS: &'static [fn(
```

Comment:
Why? I don't see any reason for doing this. This can be done entirely on the function level.

Reply:
I am explicitly provide a lifetime, because compiler can not infer it, so I am getting this error
```
error[E0310]: the parameter type `BlockNumber` may not live long enough
  --> frame/support/src/traits/hooks.rs:67:2
   |
66 |   pub trait OnIdle<BlockNumber> {
   |                    ----------- help: consider adding an explicit lifetime bound...: `BlockNumber: 'static`
67 | /     const ON_IDLE_FUNCTIONS: &'static [fn(
68 | |         BlockNumber,
69 | |         crate::weights::Weight,
70 | |     ) -> crate::weights::Weight] = &[Self::on_idle];
   | |____________________________________________________^ ...so that the reference type `&'static [fn(BlockNumber, u64) -> u64]` does not outlive the data it points at

error: aborting due to previous error

For more information about this error, try `rustc --explain E0310`.
error: could not compile `frame-support`

To learn more, run the command again with --verbose.
```

Reply:
I spoke about the const associated type. Not really required here?

Reply:
There are a few problems to do it.
As `ON_IDLE_FUNCTIONS` so we need to explicitly typed it, where we have the following type for it
```
&'static [fn(BlockNumber, crate::weights::Weight) -> crate::weights::Weight]
```
But as a BlockNumber is a generic type it is not possible to use inside this function https://doc.rust-lang.org/error-index.html#E0401 .

Reply:
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=bcfb2b0ee9438333faecef948c6da99f

Reply:
But it does not work with `consts`
https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=3db75848b9657a0649e02f9a641ad48c

Reply:
I know, but why do we need const?

Reply:
According to the documentation of the https://github.com/bkchr/impl-trait-for-tuples, it seems it is not possible to use in some case like this one, because for_tuple!() macro has some restrictions. So the following code will not compile 
```
let functions = for_tuples!( &[ #(Tuple::on_idle),* ] );
```
With the error `expected one of: `type`, `const`, parentheses, `#`, `where``.
The only how to solve it and not to use `const` is to use another custom macro for the tuple trait impl maybe something similar for this one:
```
macro_rules! tuple_impls {
    ( $( $name:ident )+ ) => {
        impl<$($name: Foo),+> Foo for ($($name,)+)
        {
            fn do_something(self) -> Self {
                let ($($name,)+) = self;
                ($($name.do_something(),)+)
            }
        }
    };
}

tuple_impls! { A }
tuple_impls! { A B }
tuple_impls! { A B C }
tuple_impls! { A B C D }
tuple_impls! { A B C D E }
tuple_impls! { A B C D E F }
tuple_impls! { A B C D E F G }
tuple_impls! { A B C D E F G H }
tuple_impls! { A B C D E F G H I }
tuple_impls! { A B C D E F G H I J }
tuple_impls! { A B C D E F G H I J K }
tuple_impls! { A B C D E F G H I J K L }
```
But I am not sure that it will be the best idea.
Or maybe you how to better use macroses from https://github.com/bkchr/impl-trait-for-tuples ?

Reply:
```
trait Trait<V> {
        fn something(val: V) -> V;
    }

    #[impl_for_tuples(1, 5)]
    impl<V: std::ops::Add<Output = V> + Default + Copy> Trait<V> for Tuple {
        fn something(val: V) -> V {
            let fns: &[fn(V) -> V] = &[for_tuples!( #( Tuple::something ),* )];
            let mut res: V = Default::default();
            for f in fns {
                res = res.add(f(val));
            }

            res
        }
    }

    struct Test;
    impl Trait<u32> for Test {
        fn something(val: u32) -> u32 {
            val * 3
        }
    }

    assert_eq!(3 * 5 * 3, <(Test, Test, Test)>::something(5));
    assert_eq!(3 * 5 * 2, <(Test, Test)>::something(5));
```

This works for me

Reply:
Yes, got it, I have not tried this combination `&[for_tuples!( #( Tuple::something ),* )];` for this `for_tuples` macro, thank you.
Now it looks much better!